huaweicloud/services/waf/resource_huaweicloud_waf_domain.go

```diff
@@ -150,170 +163,133 @@ func buildWafDomainServers(d *schema.ResourceData) []domains.ServerOpts {
 		}
 	}
 
-	logp.Printf("[DEBUG] build WAF domain ServerOpts: %#v", serverOpts)
+	log.Printf("[DEBUG] build WAF domain serverOpts: %#v", serverOpts)
 	return serverOpts
 }
 
-func resourceWafDomainV1Create(d *schema.ResourceData, meta interface{}) error {
-	config := meta.(*config.Config)
-	wafClient, err := config.WafV1Client(config.GetRegion(d))
+func flattenDomainServerAttrs(dm *domains.Domain) []map[string]interface{} {
+	servers := make([]map[string]interface{}, len(dm.Servers))
+	for i, server := range dm.Servers {
+		servers[i] = map[string]interface{}{
+			"client_protocol": server.FrontProtocol,
+			"server_protocol": server.BackProtocol,
+			"address":         server.Address,
+			"port":            server.Port,
+		}
+	}
+	return servers
+}
+
+func updateWafDomain(wafClient *golangsdk.ServiceClient, d *schema.ResourceData, cfg *config.Config) error {
+	if d.HasChanges("certificate_id", "server", "proxy") {
+		proxy := d.Get("proxy").(bool)
+
+		updateOpts := domains.UpdateOpts{
+			CertificateId:       d.Get("certificate_id").(string),
+			CertificateName:     d.Get("certificate_name").(string),
+			Servers:             buildWafDomainServers(d),
+			Proxy:               &proxy,
+			EnterpriseProjectId: cfg.GetEnterpriseProjectID(d),
+		}
+
+		log.Printf("[DEBUG] updateOpts: %#v", updateOpts)
+
+		if _, err := domains.Update(wafClient, d.Id(), updateOpts).Extract(); err != nil {
+			return fmt.Errorf("error updating WAF domain: %s", err)
+		}
+	}
+	return nil
+}
+
+func resourceWafDomainCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	wafClient, err := cfg.WafV1Client(cfg.GetRegion(d))
 	if err != nil {
-		return fmtp.Errorf("error creating HuaweiCloud WAF Client: %s", err)
+		return diag.Errorf("error creating WAF client: %s", err)
 	}
 
 	proxy := d.Get("proxy").(bool)
 
-	createOpts := domains.CreateOpts{
-		HostName:            d.Get("domain").(string),
-		CertificateId:       d.Get("certificate_id").(string),
-		CertificateName:     d.Get("certificate_name").(string),
-		Servers:             buildWafDomainServers(d),
-		Proxy:               &proxy,
-		PaidType:            d.Get("charging_mode").(string),
-		EnterpriseProjectId: config.GetEnterpriseProjectID(d),
-	}
-	logp.Printf("[DEBUG] CreateOpts: %#v", createOpts)
+	createOpts := buildCreateDomainHostOpts(d, cfg, proxy)
```

Comment:
I think we can get `proxy` from parameter `d` in function `buildCreateDomainHostOpts`. No need to define it outside the method